File: services/api-gateway/src/tracing.ts
/**
 * OpenTelemetry tracing setup for the API gateway (P6-1).
 *
 * Must be imported BEFORE any other module to ensure all HTTP/Express
 * handlers are automatically instrumented.
 *
 * Environment variables:
 *   OTEL_EXPORTER_OTLP_ENDPOINT  — e.g. "http://jaeger:4318"
 *   OTEL_SERVICE_NAME             — e.g. "api-gateway"
 *   OTEL_ENABLED                  — set to "false" to disable (default: true)
 */

import { NodeSDK } from "@opentelemetry/sdk-node";
import { getNodeAutoInstrumentations } from "@opentelemetry/auto-instrumentations-node";
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-http";
import { Resource } from "@opentelemetry/resources";
import { ATTR_SERVICE_NAME } from "@opentelemetry/semantic-conventions";

const enabled = (process.env.OTEL_ENABLED ?? "true") !== "false";
const endpoint = process.env.OTEL_EXPORTER_OTLP_ENDPOINT;

let sdk: NodeSDK | undefined;

if (enabled && endpoint) {
  const exporter = new OTLPTraceExporter({
    url: `${endpoint}/v1/traces`,
  });

  sdk = new NodeSDK({
    resource: new Resource({
      [ATTR_SERVICE_NAME]: process.env.OTEL_SERVICE_NAME ?? "api-gateway",
    }),
    traceExporter: exporter,
    instrumentations: [
      getNodeAutoInstrumentations({
        "@opentelemetry/instrumentation-fs": { enabled: false },
      }),
    ],
  });

  sdk.start();
  console.log(
    `[otel] Tracing initialised for ${process.env.OTEL_SERVICE_NAME ?? "api-gateway"} → ${endpoint}`,
  );
} else {
  console.log("[otel] Tracing disabled (OTEL_ENABLED=false or no endpoint)");
}

export function shutdownTracing(): Promise<void> {
  if (sdk) return sdk.shutdown();
  return Promise.resolve();
}

File: services/_common/_tracing.py
"""OpenTelemetry tracing initialisation — shared across all Python services.

Configures a TracerProvider with OTLP exporter and auto-instruments FastAPI.
Import and call ``init_tracing(app)`` in each service's ``main.py`` after
creating the FastAPI app.

Environment variables:
    OTEL_SERVICE_NAME      — logical service name (e.g. "flight-service")
    OTEL_EXPORTER_OTLP_ENDPOINT — OTLP HTTP endpoint (e.g. "http://jaeger:4318")
    OTEL_ENABLED           — set to "false" to disable tracing (default: true)
"""

from __future__ import annotations

import logging
import os
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from fastapi import FastAPI

logger = logging.getLogger(__name__)

_tracer_provider = None


def init_tracing(app: "FastAPI", service_name: str | None = None) -> None:
    """Initialise OpenTelemetry tracing and instrument the FastAPI app.

    Safe to call even when tracing is disabled — becomes a no-op.
    """
    global _tracer_provider

    if os.getenv("OTEL_ENABLED", "true").lower() == "false":
        logger.info("OpenTelemetry tracing disabled (OTEL_ENABLED=false)")
        return

    endpoint = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT", "")
    if not endpoint:
        logger.info("OpenTelemetry tracing disabled (no OTEL_EXPORTER_OTLP_ENDPOINT)")
        return

    try:
        from opentelemetry import trace
        from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
        from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
        from opentelemetry.sdk.resources import Resource
        from opentelemetry.sdk.trace import TracerProvider
        from opentelemetry.sdk.trace.export import BatchSpanProcessor
    except ImportError:
        logger.warning("OpenTelemetry packages not installed — tracing disabled")
        return

    name = service_name or os.getenv("OTEL_SERVICE_NAME", "unknown-service")

    resource = Resource.create({"service.name": name})
    _tracer_provider = TracerProvider(resource=resource)

    exporter = OTLPSpanExporter(endpoint=f"{endpoint}/v1/traces")
    _tracer_provider.add_span_processor(BatchSpanProcessor(exporter))

    trace.set_tracer_provider(_tracer_provider)

    FastAPIInstrumentor.instrument_app(app)

    logger.info("OpenTelemetry tracing initialised for %s → %s", name, endpoint)


def get_trace_context() -> dict[str, str]:
    """Extract current trace context as a dict for Kafka envelope injection.

    Returns a dict with ``trace_id`` and ``span_id`` keys, or empty strings
    if no active span exists.
    """
    try:
        from opentelemetry import trace

        span = trace.get_current_span()
        ctx = span.get_span_context()
        if ctx and ctx.trace_id:
            return {
                "trace_id": format(ctx.trace_id, "032x"),
                "span_id": format(ctx.span_id, "016x"),
            }
    except Exception:
        pass
    return {"trace_id": "", "span_id": ""}


def shutdown_tracing() -> None:
    """Flush and shut down the tracer provider."""
    global _tracer_provider
    if _tracer_provider:
        _tracer_provider.shutdown()
        _tracer_provider = None
